[mpris] Rewrite from mprisqt. Contributes to JB#53227

inzanity opened the pull request.

inzanity committed:
[mpris] Rewrite from mprisqt. Contributes to JB#53227

vigejolla commented on rpm/amber-mpris.spec:
```diff
@@ -0,0 +1,80 @@
+Summary:    Qt and QML MPRIS interface and adaptor
+Version:    1.0.0
+Release:    1
+License:    LGPLv2
```
Should be LGPLv2+

vigejolla commented on src/mprisplayerinterface.cpp:
```diff
@@ -2,10 +2,7 @@
- *
- * Contact: Valerio Valerio <valerio.valerio@jolla.com>
- * Author: Andres Gomez <andres.gomez@jolla.com>
+ * Copyright (C) 2015-2020 Jolla Ltd.
```
maybe 2021?

inzanity replied:
Done (and other places too)

pvuorela commented on rpm/amber-mpris.spec:
```diff
@@ -0,0 +1,80 @@
+%qmake5_install
+#mkdir -p %{buildroot}%{_libdir}/qt5/qml/org/nemomobile/mpris
+#ln -sf %{_libdir}/qt5/qml/Amber/Mpris/libambermpris-qml-plugin.so %{buildroot}%{_libdir}/qt5/qml/org/nemomobile/mpris/
+#sed 's/Amber\.Mpris/org.nemomobile.mpris/' < declarative/qmldir > %{buildroot}%{_libdir}/qt5/qml/org/nemomobile/mpris/qmldir
```
Can remove these now.

inzanity replied:
Done, also dropped the whole legacy subdir.

pvuorela commented on declarative/declarative.pro:
```diff
@@ -14,21 +14,26 @@ EXAMPLE = ../example/declarative/*
-PLUGIN_IMPORT_PATH = org/nemomobile/mpris
+PLUGIN_IMPORT_PATH = Amber/Mpris
 
 QMAKE_SUBSTITUTES = qmldir.in
```
Could get rid of this substitute and just provide a qmldir file?

inzanity replied:
Dropped.

pvuorela commented on src/src.pro:
```diff
@@ -26,40 +26,48 @@ CONFIG += create_pc create_prl no_install_prl
 headers.files = $$INSTALL_HEADERS
-headers.path = $$[QT_INSTALL_HEADERS]/MprisQt
+headers.path = /usr/include/AmberMpris
 prf.files = $${TARGET}.prf
```
Old detail, but also this having somewhat pointless substitution.

inzanity replied:
Dropped.

pvuorela commented on src/mprismetadata.h:
```diff
@@ -0,0 +1,251 @@
+
+    Q_PROPERTY(QVariant userRating READ userRating WRITE setUserRating NOTIFY metaDataChanged)
+
+    /* QDeclarativeMediaMetaData fields not in MPRIS spec
```
Myself I could just drop this sort of details. Many would be somewhat pointless exposed to external control and possible we don't ever add stuff not in spec. Different API is different API anyhow.

inzanity replied:
Yeah, I'll remove this, was just a reminder to self.

pvuorela commented on src/MprisService:
```diff
@@ -0,0 +1 @@
+#include "mprisservice.h"
```
Doesn't exist?

inzanity replied:
Oops, left over file.

pvuorela commented on src/mprisplayer.h:
```diff
@@ -53,111 +47,82 @@ class MPRIS_QT_EXPORT MprisPlayer : public QObject,
 
-    QVariantMap metadata() const;
-    void setMetadata(const QVariantMap &metadata);
+public Q_SLOTS:
```
Why are these slots? Already declared as properties.

inzanity replied:
Thanks, will remove.

pvuorela commented on declarative/declarativemprisplayer_p.h:
```diff
@@ -0,0 +1,38 @@
+
+    virtual qlonglong position() const;
+protected:
+    Q_INVOKABLE QVariant getPosition() const;
```
What's the idea on this position thing? Why not just app setting position property to something?

inzanity replied:
If the position property is set, it is set using property binding; and for optimization reasons, most player components will probably do that at reasonable intervals (likely 1s). However, should a position request come between those notifications, the returned value will be off by anything up to that interval.

The idea of the getter is to enable usage of the actual value, so the returned value will be exact. (but by default, it will return the property value, so that one can be used as well).

pvuorela replied:
The resulting API bothers me. Having a function declared with a specific name is somewhat random interface. Hard to follow what is happening when reading the code and easy to break.

1s internal could be quite ok for real-life usage already, more precise values start to suffer from d-bus delays. But if there's need for bigger update intervals or more precision, maybe there could be some positionRequested (or refreshRequested, or positionRefreshRequestd, exact name doesn't matter) signal that's emitted before reading the value?

inzanity replied:
To me that sounds like a really cumbersome solution. To me this feels natural (method overloading is quite normal OOP practice), and it is also optional (if not used, the normal property binding is used), and would be documented too.

pvuorela replied:
Normal OOP sure, but QML is not too well equipped for overloading methods, normally there it's all just properties and signals etc. Don't immediately recall if such a way is used somewhere else on public API. For c++ the player class is not either offering too much inheritability with a one single virtual method.

In practice I could expect any QML player likely has some player component with a position property it updates at least once per second. If there's such a signal emitted anyway, updating the position here on the property shouldn't be too expensive.

inzanity replied:
Tried with the signalling thing, and it does seem quite reasonable after all, happy with this?

pvuorela commented on declarative/MprisAudio.qml:
```diff
@@ -0,0 +1,136 @@
+import QtMultimedia 5.6
+import Amber.Mpris 2.0
+
+MprisPlayer {
```
I could still move this to examples. Eases maintenance and real apps likely will anyway have their own qml they develop further.

abranson replied:
It does sounds quite cool to just be able to drop this into an Audio and forget about it though. Apps doing that would avoid this boilerplate, and we'd be free to improve it for all.

inzanity replied:
(I realized that it doesn't quite work as I wanted; as Audio does not have a default property, one cannot just "drop" stuff there, and even when placing in a property, the parent item does not get set, so the actual player needs to be defined always)

inzanity replied:
The part was moved under examples/qml, maybe we can add it later on if deemed useful.

pvuorela reviewed (commented):
Added some comments, few of them being existing details I just noted.

On the API, the Manager and Controller with mostly duplicated API still confuse me, always have to separately check what was the difference, amplified by "manager" being such generic word.

Was thinking if the classes could be combined. E.g. provide list of known service in one property and have target service on another. If target is not explicitly set, then controlling the latest one as the logic now goes. 

Could be done separately but anyhow something to consider before declaring API stable.

pvuorela commented on src/mprismetadata.cpp:
```diff
@@ -0,0 +1,762 @@
+        priv->m_signalPropertyMap.clear();
+    }
+
+    priv->m_fillFromObject = qvariant_cast<QObject *>(fillFrom);
```
This would need to track also the life-cycle of the object.

inzanity replied:
m_fillFromObject is a QPointer, so the life-cycle is tracked... ...except in once place. Good catch.

inzanity replied:
Now all places should check existence before using.

abranson commented on src/MprisService:
```diff
@@ -0,0 +1 @@
+#include "mprisservice.h"
```
I don't see this header anywhere?

inzanity replied:
Yep, hopefullly removed all traces of this obsolete file.

inzanity commented:
> Was thinking if the classes could be combined. E.g. provide list of known service in one property and have target service on another. If target is not explicitly set, then controlling the latest one as the logic now goes.

This would essentially mean not exposing the MprisController at all, making it a private implementation detail; which is what the old API (on QML side) did. However, that one becomes quite hard to use, if you want to make an application that shows a list of players (essentially forcing the usage of MprisManager for each player + 1 for getting the list of players), and all of the managers would do the same discovery dance.

I'm open to naming suggestions, but I do prefer the separation as it currently stands.

abranson reviewed (approved):
I'd drop that legacy subdir, as we'll keep the old lib around for as long as it's needed. Otherwise LGTM

pvuorela commented:
> This would essentially mean not exposing the MprisController at all, making it a private implementation detail; which is what the old API (on QML side) did. However, that one becomes quite hard to use, if you want to make an application that shows a list of players (essentially forcing the usage of MprisManager for each player + 1 for getting the list of players), and all of the managers would do the same discovery dance.

Hm, true to some extend, though the discovery part can be shared in the implementation side.

Now started pondering how the current qml api should be used regarding the controllers. The controllers can't be instantiated, only the manager can give a list of objects (which it owns). The app could loop over them and do do stuff, or find a service that it sets as target of the manager, but then it's again just one component used for the actual control.

> I'm open to naming suggestions, but I do prefer the separation as it currently stands.

How about renaming manager -> controller, current controller -> serviceController or something like that?

pvuorela commented on src/mpriscontroller.h:
```diff
@@ -192,21 +196,12 @@ public Q_SLOTS:
-    MprisRootInterface *m_mprisRootInterface;
-    MprisPlayerInterface *m_mprisPlayerInterface;
+    virtual void connectNotify(const QMetaMethod &method);
+    virtual void disconnectNotify(const QMetaMethod &method);
```
Protected could be better for these two as they are overloaded methods?

inzanity replied:
Done, same for manager.

pvuorela commented on src/mprismanager.h:
```diff
@@ -26,30 +23,31 @@
 {
     Q_OBJECT
 
     Q_PROPERTY(bool singleService READ singleService WRITE setSingleService NOTIFY singleServiceChanged)
```
Existing thing, but this seems an interesting property and not entirely sure how to interpret it. The app developer set separately set currentService and singleService but what do true or false on the latter mean...

inzanity replied:
Yeah, I'm not entirely sure whether simpleService makes that much sense if the controllers are exposed (apart from the lifetime things).

If singleService is not set, then setting currentService doesn't have very much effect, as it can change to something else at a moment's notice.

inzanity replied:
I'm not entirely sure what to do here; if singleService is removed, then currentService should be set read only; but this has a certain problem that as client's can't be instantiated, wanting to control specific player would become quite hard.

Another issue is that currently MprisPlayer appends a ".instance<pid>" to the service name, so even with the singleService  & currentService, it might not be possible to control the same player across restarts.

pvuorela commented on src/mpriscontroller.cpp:
```diff
@@ -26,58 +23,135 @@
+
+void MprisControllerPrivate::onPositionTimeout()
+{
+    if (QDateTime::currentMSecsSinceEpoch() - m_lastStamp > m_syncInterval) {
```
How about if the time gets changed? QElapsedTimer could be more robust?

inzanity replied:
It might also be a bit simpler, let's give it a go.

inzanity replied:
Changed to QElapsedTimer.

pvuorela commented on src/mpriscontroller.cpp:
```diff
@@ -152,10 +212,10 @@ bool MprisController::next()
-    QObject::connect(watcher, SIGNAL(finished(QDBusPendingCallWatcher*)),
-                     this, SLOT(onFinishedPendingCall(QDBusPendingCallWatcher*)));
+    QObject::connect(watcher, &QDBusPendingCallWatcher::finished,
+                     priv, &MprisControllerPrivate::onFinishedPendingCall);
```
Old thing, but these three lines get repeated a bunch of times. Could combine to some helper method.

inzanity replied:
Yeah, true, no reason to prevent inheritants from overloading them.

inzanity replied:
Done.

pvuorela commented on src/mpriscontroller.cpp:
```diff
@@ -187,11 +247,11 @@ bool MprisController::openUri(const QUrl &uri)
     for (int i = 0; i < mimeNames.size(); i++) {
-        if (m_mprisRootInterface->supportedMimeTypes().contains(mimeNames[i])) {
-            QDBusPendingReply<> reply = m_mprisPlayerInterface->OpenUri(uri.toString());
+        if (priv->m_mprisRootInterface.supportedMimeTypes().contains(mimeNames[i])) {
```
Old thing too, but that mimetype guessing and filtering seems pointless btw. The client can surely figure out itself if it's able to open something.

inzanity replied:
Agreed, both the mime type and scheme validation should probably be dropped; I expect many players to be able to handle things they don't even know they can, so the supported lists are probably not very relevant.

inzanity replied:
Dropped.

inzanity commented:
One thing I noticed during reviewed code testing, is that one can not override the fillFrom on a per-field basis, which would sound quite useful. It could be changed so that explicitly set properties are preferred, and "fillFrom" ones are only used.

This was mainly issue with desktop qtmultimedia, which does not have duration defined in metaData, but it gets always read from there.

This would be a non API breaking change though, so it could be changed later on too.

pvuorela commented on common.pri:
```diff
@@ -1 +1,2 @@
-MPRISQTLIB = mpris-qt5
+MPRISQTLIB = ambermpris
+CONFIG += debug
```
This shouldn't be here?

inzanity replied:
True, leftover from crash investigation

inzanity commented:
> Hm, true to some extend, though the discovery part can be shared in the implementation side.

Yeah, sharing a singleton management entity would work quite nicely.

> Now started pondering how the current qml api should be used regarding the controllers. The controllers can't be instantiated, only the manager can give a list of objects (which it owns). The app could loop over them and do do stuff, or find a service that it sets as target of the manager, but then it's again just one component used for the actual control.

My main idea for publishing the list of controllers is that the service name in itself is not very useful, and building a list of something by looping through the names is quite cumbersome. (which was especially true in the old version, which advertised the service names before properties had been fetched; and also changed the order of the list when a service was selected).

> > I'm open to naming suggestions, but I do prefer the separation as it currently stands.
> 
> How about renaming manager -> controller, current controller -> serviceController or something like that?

How about client instead of serviceController?

pvuorela commented:
> My main idea for publishing the list of controllers is that the service name in itself is not very useful, and building a list of something by looping through the names is quite cumbersome. (which was especially true in the old version, which advertised the service names before properties had been fetched; and also changed the order of the list when a service was selected).

Hm, if we'd need to support UI that presented a list of mpris players, perhaps there should be a model that exposed the information. Either just some plain one that had icon, desktop file, title, and the app could instantiate a controller separately if wanted. Or then more full-blown version like the one here.

But that's maybe not needed now.

> How about client instead of serviceController?

So Controller (that takes care of the current target) and Client (for a specific one)? Sounds fine too.

pvuorela commented:
Getting quite nice! Guess only needing some notes on what parts are considered internal at this point, i.e. the controller side and then the fillFrom property. 

For extra detail that singleService still feels strange and not sure if anything actually needs it. Dropping could be an option or marking it even more internal, but then again if the controller is internal now, it shouldn't matter too much.

Some commented out properties were removed from metadata. There's still commented out method stubs in .h/.cpp, wonder if intentional.

Phew.

pvuorela commented:
Actually the build fails, .spec still has MprisManager etc.

inzanity commented:
> Getting quite nice! Guess only needing some notes on what parts are considered internal at this point, i.e. the controller side and then the fillFrom property.

The internal markings are already present in the jb56036 branch with the documentation; I could pull in the docs to the same PR though.

> For extra detail that singleService still feels strange and not sure if anything actually needs it. Dropping could be an option or marking it even more internal, but then again if the controller is internal now, it shouldn't matter too much.

I'm a little hesitant on dropping the singleService, as without it the controller cannot be used at all to control a specific instance; not probably too relevant on our platform though.

> Some commented out properties were removed from metadata. There's still commented out method stubs in .h/.cpp, wonder if intentional.

Thanks, removed.

> Actually the build fails, .spec still has MprisManager etc.

Thanks, fixed.

pvuorela commented:
> The internal markings are already present in the jb56036 branch with the documentation; I could pull in the docs to the same PR though.

Good good. I'm ok, and maybe even preferring doing separately to avoid going through all these changes again and again :)

> I'm a little hesitant on dropping the singleService

Fine as well. As long as we don't make promises.

Now with the .spec build failure adjusted, I have no more objections on getting this in.
-> LGTM.
